Packet++/header/S7commLayer.h

```diff
@@ -0,0 +1,234 @@
+		/** data length */
+		uint16_t data_length;
+	} s7commhdr;
+#pragma pack(pop)
+
+/**
+ * @struct s7comm_ack_data_hdr
+ * Represents a S7COMM protocol with Ack-Data header
+ */
+#pragma pack(push, 1)
+	struct s7comm_ack_data_hdr : s7commhdr
+	{
+		/** error class */
+		uint8_t error_class;
+		/** error code */
+		uint8_t error_code;
+	};
+#pragma pack(pop)
+
+	/**
+	 * @class S7CommParameter
+	 * Represents a S7COMM (S7 Communication7) protocol Parameter
+	 */
+	class S7CommParameter
+	{
+		friend class S7commLayer;
+
+	  public:
+		S7CommParameter() {}
+
+		virtual ~S7CommParameter() {}
+
+		/**
+		 * @return The data of the Parameter
+		 */
+		uint8_t *getData() const { return m_Data; }
+		/**
+		 * @return The length of the Parameter data
+		 */
+		size_t getDataLength() const { return m_DataLen; }
+
+	  private:
+		S7CommParameter(uint8_t *data, size_t dataLen) : m_Data(data), m_DataLen(dataLen) {}
+		uint8_t *m_Data;
+		size_t m_DataLen;
+	};
+	/**
+	 * @class S7commLayer
+	 * Represents a S7COMM (S7 Communication) protocol
+	 */
+	class S7commLayer : public Layer
+	{
+	  public:
+		/**
+		 * A constructor that allocates a new S7comm header
+		 * @param[in] msg_type The general type of the message
+		 * @param[in] pdu_ref Link responses to their requests
+		 * @param[in] param_length The length of the parameter field
+		 * @param[in] data_length The length of the data field
+		 * @param[in] error_class The value of the error class
+		 * @param[in] error_code The value of the error code
+		 */
+		S7commLayer(uint8_t msg_type, uint16_t pdu_ref, uint16_t param_length, uint16_t data_length,
+					uint8_t error_class = 0, uint8_t error_code = 0);
+
+		/**
+		 * A constructor that creates the layer from an existing packet raw data
+		 * @param[in] data A pointer to the raw data (will be casted to @ref s7commhdr)
+		 * @param[in] dataLen Size of the data in bytes
+		 * @param[in] prevLayer A pointer to the previous layer
+		 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+		 */
+		S7commLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+			: Layer(data, dataLen, prevLayer, packet)
+		{
+			m_Protocol = S7COMM;
+			m_Parameter = nullptr;
+		}
+
+		virtual ~S7commLayer() { delete m_Parameter; }
+
+		/**
+		 * @return S7comm protocol id
+		 */
+		uint8_t getProtocolId() const;
+
+		/**
+		 * @return S7comm message type
+		 */
+		uint8_t getMsgType() const;
+
+		/**
+		 * @return S7comm reserved
+		 */
+		uint16_t getReserved() const;
+
+		/**
+		 * @return S7comm PDU ref
+		 */
+		uint16_t getPduRef() const;
+
+		/**
+		 * @return S7comm parameter length
+		 */
+		uint16_t getParamLength() const;
+
+		/**
+		 * @return S7comm data length
+		 */
+		uint16_t getDataLength() const;
+
+		/**
+		 * @return S7comm error code
+		 */
+		uint8_t getErrorCode() const;
+		/**
+		 * @return S7comm error class
+		 */
+		uint8_t getErrorClass() const;
+
+		const S7CommParameter *getParameter();
```

Comment:
the signature should be:
```cpp
const S7CommParameter* getParameter() const;
```

Also - please add doxygen documentation?

Reply:
I'm not sure it can be const. In the S7Comm.cpp file, I give m_Paramer a value.

What do you think?

Reply:
Ok sure 👍